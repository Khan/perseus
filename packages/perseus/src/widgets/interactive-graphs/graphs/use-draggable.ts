import {useDrag} from "@use-gesture/react";
import {useTransformContext, vec} from "mafs";
import * as React from "react";
import invariant from "tiny-invariant";

import {X, Y} from "../math";
import useGraphConfig from "../reducer/use-graph-config";

import type {RefObject} from "react";

/**
 * Code in this file is derived from
 * https://github.com/stevenpetryk/mafs/blob/4520319379a2cc2df8148d8baaef1f85db117103/src/interaction/useMovable.tsx#L20-L83
 * and copied here under the terms of the MIT license.
 *
 * Copyright 2021 Steven Petryk
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 */

export type Params = {
    gestureTarget: RefObject<Element>;
    onMove: (point: vec.Vector2) => unknown;
    point: vec.Vector2;
    constrain: (point: vec.Vector2) => vec.Vector2;
};

type DragState = {
    dragging: boolean;
};

export function useDraggable(args: Params): DragState {
    const {gestureTarget: target, onMove, point, constrain} = args;
    const [dragging, setDragging] = React.useState(false);
    const {xSpan, ySpan} = useSpanContext();
    const {viewTransform, userTransform} = useTransformContext();

    const inverseViewTransform = vec.matrixInvert(viewTransform);
    invariant(inverseViewTransform, "The view transform must be invertible.");

    const inverseTransform = React.useMemo(
        () => getInverseTransform(userTransform),
        [userTransform],
    );

    const pickup = React.useRef<vec.Vector2>([0, 0]);

    useDrag(
        (state) => {
            const {type, event} = state;
            event?.stopPropagation();

            const isKeyboard = type.includes("key");
            if (isKeyboard) {
                event?.preventDefault();

                // When a key is held down, we see multiple "keydown" events,
                // followed by a single "keyup" event.
                // For a single keypress, we only see a "keydown" event, no
                // "keyup".
                // We never want to process the keyup event as an intent to
                // move so we bail on further processing here.
                if (type === "keyup") {
                    return;
                }

                const {
                    direction: yDownDirection,
                    altKey,
                    metaKey,
                    shiftKey,
                } = state;

                const direction = [
                    yDownDirection[X],
                    -yDownDirection[Y],
                ] as vec.Vector2;
                const span = Math.abs(direction[X]) ? xSpan : ySpan;

                let divisions = 50;
                if (altKey || metaKey) {
                    divisions = 200;
                }
                if (shiftKey) {
                    divisions = 10;
                }

                const min = span / (divisions * 2);
                const tests = range(
                    span / divisions,
                    span / 2,
                    span / divisions,
                );

                for (const dx of tests) {
                    // Transform the test back into the point's coordinate system
                    const testMovement = vec.scale(direction, dx);
                    const testPoint = constrain(
                        vec.transform(
                            vec.add(
                                vec.transform(point, userTransform),
                                testMovement,
                            ),
                            inverseTransform,
                        ),
                    );

                    if (vec.dist(testPoint, point) > min) {
                        onMove(testPoint);
                        break;
                    }
                }
            } else {
                const {last, movement: pixelMovement, first} = state;

                setDragging(!last);

                if (first) {
                    pickup.current = vec.transform(point, userTransform);
                }
                if (vec.mag(pixelMovement) === 0) {
                    return;
                }

                const movement = vec.transform(
                    pixelMovement,
                    inverseViewTransform,
                );
                onMove(
                    constrain(
                        vec.transform(
                            vec.add(pickup.current, movement),
                            inverseTransform,
                        ),
                    ),
                );
            }
        },
        {target, eventOptions: {passive: false}},
    );
    return {dragging};
}

function getInverseTransform(transform: vec.Matrix) {
    const invert = vec.matrixInvert(transform);
    invariant(
        invert !== null,
        "Could not invert transform matrix. A parent transformation matrix might be degenerative (mapping 2D space to a line).",
    );
    return invert;
}

function useSpanContext() {
    const {
        range: [[xMin, xMax], [yMin, yMax]],
    } = useGraphConfig();
    const xSpan = xMax - xMin;
    const ySpan = yMax - yMin;
    return {xSpan, ySpan};
}

function range(min: number, max: number, step = 1): number[] {
    const result: number[] = [];
    for (let i = min; i < max - step / 2; i += step) {
        result.push(i);
    }

    const computedMax = result[result.length - 1] + step;
    if (Math.abs(max - computedMax) < step / 1e-6) {
        result.push(max);
    } else {
        result.push(computedMax);
    }

    return result;
}
