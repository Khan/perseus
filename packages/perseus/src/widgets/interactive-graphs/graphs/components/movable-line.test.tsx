import {render} from "@testing-library/react";
import {Mafs} from "mafs";
import React from "react";

import * as UseDraggableModule from "../use-draggable";

import {
    getMovableLineKeyboardConstraint,
    MovableLine,
    trimRange,
} from "./movable-line";

import type {Interval, vec} from "mafs";

describe("trimRange", () => {
    it("does not trim ranges below a size of 0", () => {
        const graphDimensionsInPixels: vec.Vector2 = [1, 1];
        const range: [Interval, Interval] = [
            [0, 1],
            [0, 1],
        ];

        const trimmed = trimRange(range, graphDimensionsInPixels);

        expect(trimmed).toEqual([
            [0.5, 0.5],
            [0.5, 0.5],
        ]);
    });

    it("trims 4 units from each edge when a unit is one pixel", () => {
        const graphDimensionsInPixels: vec.Vector2 = [20, 20];
        const range: [Interval, Interval] = [
            [-10, 10],
            [-10, 10],
        ];

        const trimmed = trimRange(range, graphDimensionsInPixels);

        expect(trimmed).toEqual([
            [-6, 6],
            [-6, 6],
        ]);
    });

    it("trims 0.4 units from each edge when a unit is ten pixels", () => {
        const graphDimensionsInPixels: vec.Vector2 = [200, 200];
        const range: [Interval, Interval] = [
            [-10, 10],
            [-10, 10],
        ];

        const trimmed = trimRange(range, graphDimensionsInPixels);

        expect(trimmed).toEqual([
            [-9.6, 9.6],
            [-9.6, 9.6],
        ]);
    });

    it("doesn't mix up x and y", () => {
        const graphDimensionsInPixels: vec.Vector2 = [200, 400];
        const range: [Interval, Interval] = [
            [-10, 10],
            [-1, 1],
        ];

        const trimmed = trimRange(range, graphDimensionsInPixels);

        expect(trimmed).toEqual([
            [-9.6, 9.6],
            [-0.98, 0.98],
        ]);
    });
});

describe("Rendering", () => {
    let useDraggable: jest.SpyInstance;

    beforeEach(() => {
        useDraggable = jest
            .spyOn(UseDraggableModule, "useDraggable")
            .mockReturnValue({dragging: false});
    });

    it("Does NOT render extensions of line when option is not provided", () => {
        const {container} = render(
            <Mafs width={200} height={200}>
                <MovableLine
                    points={[
                        [-1, -1],
                        [1, 1],
                    ]}
                />
            </Mafs>,
        );
        expect(container).toMatchSnapshot();
    });

    it("Does NOT render extensions of line when option is disabled", () => {
        const {container} = render(
            <Mafs width={200} height={200}>
                <MovableLine
                    points={[
                        [-1, -1],
                        [1, 1],
                    ]}
                    extend={{
                        start: false,
                        end: false,
                    }}
                />
            </Mafs>,
        );
        expect(container).toMatchSnapshot();
    });

    it("Does render extensions of line when option is enabled", () => {
        const {container} = render(
            <Mafs width={200} height={200}>
                <MovableLine
                    points={[
                        [-1, -1],
                        [1, 1],
                    ]}
                    extend={{
                        start: true,
                        end: true,
                    }}
                />
            </Mafs>,
        );
        expect(container).toMatchSnapshot();
    });

    it("Adds styling when the line is being dragged", () => {
        // Verify normal non-dragging state
        let container = render(
            <Mafs width={200} height={200}>
                <MovableLine
                    points={[
                        [-1, -1],
                        [1, 1],
                    ]}
                />
            </Mafs>,
        ).container;
        // eslint-disable-next-line testing-library/no-container,testing-library/no-node-access
        let line = container.querySelector(
            ".movable-line",
        ) as HTMLElement | null;
        expect(line?.getAttribute("style")).toContain("cursor: grab;");
        // eslint-disable-next-line testing-library/no-container,testing-library/no-node-access
        line = container.querySelector("[data-testid='movable-line__line']");
        expect(line?.classList).not.toContain("movable-dragging");

        // Verify dragging state
        useDraggable.mockReturnValue({dragging: true});
        container = render(
            <Mafs width={200} height={200}>
                <MovableLine
                    points={[
                        [-1, -1],
                        [1, 1],
                    ]}
                />
            </Mafs>,
        ).container;
        // eslint-disable-next-line testing-library/no-container,testing-library/no-node-access
        line = container.querySelector(".movable-line") as HTMLElement | null;
        expect(line?.getAttribute("style")).toContain("cursor: grabbing;");
        // eslint-disable-next-line testing-library/no-container,testing-library/no-node-access
        line = container.querySelector("[data-testid='movable-line__line']");
        expect(line?.classList).toContain("movable-dragging");
    });
});
describe("getMovableLineKeyboardConstraint", () => {
    it("should snap to the snapStep and avoid putting points on a vertical line", () => {
        const line: [vec.Vector2, vec.Vector2] = [
            [0, 0],
            [1, 0],
        ];
        const snapStep: vec.Vector2 = [1, 1];

        // We're moving the first point
        const constraint = getMovableLineKeyboardConstraint(line, snapStep, 0);

        expect(constraint).toEqual({
            up: [0, 1],
            down: [0, -1],
            left: [-1, 0],
            right: [2, 0], // Avoids overlapping the points
        });
    });
});
