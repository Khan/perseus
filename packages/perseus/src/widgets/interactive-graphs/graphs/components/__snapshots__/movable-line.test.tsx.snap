// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering Does NOT render extensions of line when option is disabled 1`] = `
<div>
  <div
    class="MafsView"
    style="width: 200px; height: 200px;"
    tabindex="0"
  >
    <svg
      height="200"
      preserveAspectRatio="xMidYMin"
      style="width: 200px; --mafs-view-transform: matrix(28.57143, 0, 0, -28.57143, 0, 0); --mafs-user-transform: translate(0, 0);"
      viewBox="-100 -100 200 200"
      width="200"
    >
      <g
        aria-label="Point 1 at -1 comma -1"
        aria-live="polite"
        class="movable-point__focusable-handle"
        data-testid="movable-point__focusable-handle"
        role="button"
        tabindex="0"
      />
      <g
        class="movable-line"
        data-testid="movable-line"
        style="cursor: grab;"
        tabindex="0"
      >
        <line
          style="stroke: transparent; stroke-width: 44;"
          x1="0"
          x2="0"
          y1="0"
          y2="0"
        />
        <line
          class="movable-line-focus-outline"
          x1="0"
          x2="0"
          y1="0"
          y2="0"
        />
        <line
          class="movable-line-focus-outline-gap"
          x1="0"
          x2="0"
          y1="0"
          y2="0"
        />
        <line
          class=""
          data-testid="movable-line__line"
          style="stroke: var(--movable-line-stroke-color); stroke-width: var(--movable-line-stroke-weight);"
          x1="0"
          x2="0"
          y1="0"
          y2="0"
        />
      </g>
      <g
        aria-label="Point 2 at 1 comma 1"
        aria-live="polite"
        class="movable-point__focusable-handle"
        data-testid="movable-point__focusable-handle"
        role="button"
        tabindex="0"
      />
      <g
        class="movable-point"
        data-testid="movable-point"
        style="--movable-point-color: #1865f2;"
      >
        <circle
          class="movable-point-hitbox"
          cx="0"
          cy="0"
          r="24"
        />
        <circle
          class="movable-point-halo"
          cx="0"
          cy="0"
        />
        <circle
          class="movable-point-ring"
          cx="0"
          cy="0"
        />
        <circle
          class="movable-point-focus-outline"
          cx="0"
          cy="0"
        />
        <circle
          class="movable-point-center"
          cx="0"
          cy="0"
          data-testid="movable-point__center"
          style="fill: #1865f2;"
        />
      </g>
      <g
        class="movable-point"
        data-testid="movable-point"
        style="--movable-point-color: #1865f2;"
      >
        <circle
          class="movable-point-hitbox"
          cx="0"
          cy="0"
          r="24"
        />
        <circle
          class="movable-point-halo"
          cx="0"
          cy="0"
        />
        <circle
          class="movable-point-ring"
          cx="0"
          cy="0"
        />
        <circle
          class="movable-point-focus-outline"
          cx="0"
          cy="0"
        />
        <circle
          class="movable-point-center"
          cx="0"
          cy="0"
          data-testid="movable-point__center"
          style="fill: #1865f2;"
        />
      </g>
    </svg>
  </div>
</div>
`;

exports[`Rendering Does NOT render extensions of line when option is not provided 1`] = `
<div>
  <div
    class="MafsView"
    style="width: 200px; height: 200px;"
    tabindex="0"
  >
    <svg
      height="200"
      preserveAspectRatio="xMidYMin"
      style="width: 200px; --mafs-view-transform: matrix(28.57143, 0, 0, -28.57143, 0, 0); --mafs-user-transform: translate(0, 0);"
      viewBox="-100 -100 200 200"
      width="200"
    >
      <g
        aria-label="Point 1 at -1 comma -1"
        aria-live="polite"
        class="movable-point__focusable-handle"
        data-testid="movable-point__focusable-handle"
        role="button"
        tabindex="0"
      />
      <g
        class="movable-line"
        data-testid="movable-line"
        style="cursor: grab;"
        tabindex="0"
      >
        <line
          style="stroke: transparent; stroke-width: 44;"
          x1="0"
          x2="0"
          y1="0"
          y2="0"
        />
        <line
          class="movable-line-focus-outline"
          x1="0"
          x2="0"
          y1="0"
          y2="0"
        />
        <line
          class="movable-line-focus-outline-gap"
          x1="0"
          x2="0"
          y1="0"
          y2="0"
        />
        <line
          class=""
          data-testid="movable-line__line"
          style="stroke: var(--movable-line-stroke-color); stroke-width: var(--movable-line-stroke-weight);"
          x1="0"
          x2="0"
          y1="0"
          y2="0"
        />
      </g>
      <g
        aria-label="Point 2 at 1 comma 1"
        aria-live="polite"
        class="movable-point__focusable-handle"
        data-testid="movable-point__focusable-handle"
        role="button"
        tabindex="0"
      />
      <g
        class="movable-point"
        data-testid="movable-point"
        style="--movable-point-color: #1865f2;"
      >
        <circle
          class="movable-point-hitbox"
          cx="0"
          cy="0"
          r="24"
        />
        <circle
          class="movable-point-halo"
          cx="0"
          cy="0"
        />
        <circle
          class="movable-point-ring"
          cx="0"
          cy="0"
        />
        <circle
          class="movable-point-focus-outline"
          cx="0"
          cy="0"
        />
        <circle
          class="movable-point-center"
          cx="0"
          cy="0"
          data-testid="movable-point__center"
          style="fill: #1865f2;"
        />
      </g>
      <g
        class="movable-point"
        data-testid="movable-point"
        style="--movable-point-color: #1865f2;"
      >
        <circle
          class="movable-point-hitbox"
          cx="0"
          cy="0"
          r="24"
        />
        <circle
          class="movable-point-halo"
          cx="0"
          cy="0"
        />
        <circle
          class="movable-point-ring"
          cx="0"
          cy="0"
        />
        <circle
          class="movable-point-focus-outline"
          cx="0"
          cy="0"
        />
        <circle
          class="movable-point-center"
          cx="0"
          cy="0"
          data-testid="movable-point__center"
          style="fill: #1865f2;"
        />
      </g>
    </svg>
  </div>
</div>
`;

exports[`Rendering Does render extensions of line when option is enabled 1`] = `
<div>
  <div
    class="MafsView"
    style="width: 200px; height: 200px;"
    tabindex="0"
  >
    <svg
      height="200"
      preserveAspectRatio="xMidYMin"
      style="width: 200px; --mafs-view-transform: matrix(28.57143, 0, 0, -28.57143, 0, 0); --mafs-user-transform: translate(0, 0);"
      viewBox="-100 -100 200 200"
      width="200"
    >
      <g
        aria-label="Point 1 at -1 comma -1"
        aria-live="polite"
        class="movable-point__focusable-handle"
        data-testid="movable-point__focusable-handle"
        role="button"
        tabindex="0"
      />
      <g
        class="movable-line"
        data-testid="movable-line"
        style="cursor: grab;"
        tabindex="0"
      >
        <line
          style="stroke: transparent; stroke-width: 44;"
          x1="0"
          x2="0"
          y1="0"
          y2="0"
        />
        <line
          class="movable-line-focus-outline"
          x1="0"
          x2="0"
          y1="0"
          y2="0"
        />
        <line
          class="movable-line-focus-outline-gap"
          x1="0"
          x2="0"
          y1="0"
          y2="0"
        />
        <line
          class=""
          data-testid="movable-line__line"
          style="stroke: var(--movable-line-stroke-color); stroke-width: var(--movable-line-stroke-weight);"
          x1="0"
          x2="0"
          y1="0"
          y2="0"
        />
      </g>
      <g
        style="stroke: var(--movable-line-stroke-color); stroke-width: 2;"
      >
        <line
          x1="0"
          x2="0"
          y1="0"
          y2="0"
        />
        <g
          class="interactive-graph-arrowhead"
          transform="translate(0 0) rotate(0)"
        >
          <g
            transform="translate(-1.5)"
          >
            <path
              d="M-4.199999999999999 5.6C-3.8499999999999996 3.5 0 0.35 1.0499999999999998 0C0 -0.35 -3.8499999999999996 -3.5 -4.199999999999999 -5.6"
              fill="none"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2px"
              style="stroke: var(--movable-line-stroke-color);"
            />
          </g>
        </g>
      </g>
      <g
        style="stroke: var(--movable-line-stroke-color); stroke-width: 2;"
      >
        <line
          x1="0"
          x2="0"
          y1="0"
          y2="0"
        />
        <g
          class="interactive-graph-arrowhead"
          transform="translate(0 0) rotate(0)"
        >
          <g
            transform="translate(-1.5)"
          >
            <path
              d="M-4.199999999999999 5.6C-3.8499999999999996 3.5 0 0.35 1.0499999999999998 0C0 -0.35 -3.8499999999999996 -3.5 -4.199999999999999 -5.6"
              fill="none"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2px"
              style="stroke: var(--movable-line-stroke-color);"
            />
          </g>
        </g>
      </g>
      <g
        aria-label="Point 2 at 1 comma 1"
        aria-live="polite"
        class="movable-point__focusable-handle"
        data-testid="movable-point__focusable-handle"
        role="button"
        tabindex="0"
      />
      <g
        class="movable-point"
        data-testid="movable-point"
        style="--movable-point-color: #1865f2;"
      >
        <circle
          class="movable-point-hitbox"
          cx="0"
          cy="0"
          r="24"
        />
        <circle
          class="movable-point-halo"
          cx="0"
          cy="0"
        />
        <circle
          class="movable-point-ring"
          cx="0"
          cy="0"
        />
        <circle
          class="movable-point-focus-outline"
          cx="0"
          cy="0"
        />
        <circle
          class="movable-point-center"
          cx="0"
          cy="0"
          data-testid="movable-point__center"
          style="fill: #1865f2;"
        />
      </g>
      <g
        class="movable-point"
        data-testid="movable-point"
        style="--movable-point-color: #1865f2;"
      >
        <circle
          class="movable-point-hitbox"
          cx="0"
          cy="0"
          r="24"
        />
        <circle
          class="movable-point-halo"
          cx="0"
          cy="0"
        />
        <circle
          class="movable-point-ring"
          cx="0"
          cy="0"
        />
        <circle
          class="movable-point-focus-outline"
          cx="0"
          cy="0"
        />
        <circle
          class="movable-point-center"
          cx="0"
          cy="0"
          data-testid="movable-point__center"
          style="fill: #1865f2;"
        />
      </g>
    </svg>
  </div>
</div>
`;
